src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs

```diff
@@ -250,8 +253,46 @@ public bool MoveNext()
                         return true;
                     }
 
-                    if (_returnedResults >= _query.PageSize)
-                        return false;
+                    _skippedResults.Value++;
+                }
+
+                return false;
+            }
+
+            private void CountDocumentsInEnumerator()
+            {
+                // If we know how many documents we have, and we do not have a DISTINCT clause, we can simply return the current value from memory.
+                if (_totalResultsCalculated && _query.Metadata.IsDistinct == false)
+                    return;
+                
+                // We need to disable calculating totalResults in the enumerator, so let's set _totalResultsCalculated to true.
```

Comment:
Why are we counting manually? If we don't have distinct, can't we get that from the number of items without materializing everything?

Reply:
If we don't have distinct  clause and we know from storage how many items we have we are escaping materializing (line #265), however if we don't know how many items our enumerable has (this is for example `In`/`startsWith`) we've to go through all data. (We handled it before very similarly: https://github.com/ravendb/ravendb/blob/f632fa9ff761aa5d206eba57abe66ce48ffffc7a/src/Raven.Server/Documents/Queries/CollectionQueryEnumerable.cs#L353-L359)

However, it still seems to be not correct for cases when we've filter. For such case we've to materialize, check via script and then count, right?

Reply:
With `filter`, that require to actually execute it all, I think